packages/apollo-gateway/src/executeQueryPlan.ts

```diff
@@ -93,44 +108,87 @@ export async function executeQueryPlan<TContext>(
   return errors.length === 0 ? { data } : { errors, data };
 }
 
+// Note: this function always returns a protobuf QueryPlanNode tree, even if
+// we're going to ignore it, because it makes the code much simpler and more
+// typesafe. However, it doesn't actually ask for traces from the backend
+// service unless we are capturing traces for Engine.
 async function executeNode<TContext>(
   context: ExecutionContext<TContext>,
   node: PlanNode,
   results: ResultMap | ResultMap[],
   path: ResponsePath,
-): Promise<void> {
+  captureTraces: boolean,
+): Promise<Trace.QueryPlanNode> {
   if (!results) {
-    return;
+    // XXX I don't understand `results` threading well enough to understand when this happens
```

Comment:
So I think this is how, among other things, if you're running `query { fooOrNullFromServiceA { somethingFromServiceB } }` and the first thing returns null, then you don't run the second query at all.